crates/bevy_ecs/src/system/query.rs

```diff
@@ -543,14 +551,17 @@ where
             >())),
         }
     }
+}
 
+impl<'w, Q: WorldQuery, F: WorldQuery> Query<'w, Q, F>
+where
+    F::Fetch: FilterFetch,
+    Q::ReadOnlyFetch: Fetch<'w, State = Q::State>,
+{
     /// Returns true if this query contains no elements.
     #[inline]
     pub fn is_empty(&self) -> bool {
-        // TODO: This code can be replaced with `self.iter().next().is_none()` if/when
-        // we sort out how to convert "write" queries to "read" queries.
-        self.state
-            .is_empty(self.world, self.last_change_tick, self.change_tick)
+        self.iter().next().is_none()
```

Comment:
If this PR lands, I'm curious if we actually want to replace the implementation of `is_empty` with this. 
Now that I think of it, having a specialized `is_empty` functionality actually gives us a small performance win (even if it is at the cost of extra code duplication). Since in the implementation we never have to touch `self.fetch` so we bypass that extra unnecessary code path. 

Where as in here, we'll have to follow the normal code path and do a bit of extra work.

Reply:
I refactored this to use a `NopFetch` to get the performance benefit without the code duplication. How does that look?